### PR TITLE
Resolve regression for large payloads deploy

### LINKED ISSUE #13233: Regression in deploying large payloads
We've introduced a regression on deploying large payloads on multi-partition clusters with
- #11661 

Before that pull request, a deployed resource was written in two follow-up events:
- `Deployment:CREATED`
- `Process:CREATED` (or `DecisionRequirements:CREATED` depending on resource)

The specialized `DeploymentDistribution` would take the resource from the `Deployment:CREATED` to distribute it to the other partitions.

With #11456, a new event `CommandDistribution:STARTED` was introduced to store the command for distribution. For deployments, this contains the entire Deployment incl. the resource. This event is only appended on multi-partition clusters. But, when it is appended, it further reduces the maximum payload size because the follow-up events reach the `MAX_BATCH_SIZE` restriction with a lower payload size.

> **Note** The `MAX_BATCH_SIZE` is a limitation that originated from the `MAX_MESSAGE_SIZE` configuration setting but nowadays is only defined by the `LogStreamBuilder`'s `MAX_FRAGMENT_SIZE`: 4MB.

This regression lowers the maximum payload size of deployments from ~2MB down to ~1.4MB.

The regression does not exist on single partition clusters as these do not require the distribution of the deployment.

---

Suggested solution:
- don't write the resource in the `Deployment:CREATED` event, only in the `Process:CREATED` and the `DecisionRequirements:CREATED` events (estimate: x-small)

Of course, we'd need to inform users that we're no longer writing the resource for that event in the Update Guide. As this would be breaking user space. The documentation should clarify that the resource is available in the `Process:CREATED` and the `DecisionRequirements:CREATED` instead.

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

Removed resource from `Deployment:CREATED` event. Resource is still available in `Process:CREATED` or `DecisionRequirements:CREATED`

## Related issues

closes #13233 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java [hunk 1]
```diff
@@ -193,6 +193,21 @@ private void createDmnResources(
                 stateWriter.appendFollowUpEvent(command.getKey(), DecisionIntent.CREATED, record));
   }
 
+  /**
+   * Create a copy of the provided deployment record and write the Deployment:CREATED event with
+   * that copy
+   *
+   * @param command the already distributed command to process
+   * @param deploymentEvent the record to clone and modify
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:203]: >and modify

🔧 This reads as if the parameter may be modified as well, but only the copy is modified

**Review by korthout — CHANGES_REQUESTED** [review]
Nice work @nicpuppa 👏 

👍 I like your approach to removing the resources!

👍 Nice work on improving the test cases

💭 Seems that there were more changes involved than we originally estimated for it. Especially getting the tests right was harder than expected

❌ I have two change requests. Please have a look below

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java [hunk 2]
```diff
@@ -145,8 +145,10 @@ private void transformAndDistributeDeployment(final TypedRecord<DeploymentRecord
     }
 
     final long key = keyGenerator.nextKey();
-    responseWriter.writeEventOnCommand(key, DeploymentIntent.CREATED, deploymentEvent, command);
-    stateWriter.appendFollowUpEvent(key, DeploymentIntent.CREATED, deploymentEvent);
+    final var recordWithoutResource = createDeploymentWithoutResources(deploymentEvent);
+    responseWriter.writeEventOnCommand(
+        key, DeploymentIntent.CREATED, recordWithoutResource, command);
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:150]: Seems the response is [builded](https://github.com/camunda/zeebe/blob/e7965416755a1fac14129b60889f6d22d260fd19/gateway/src/main/java/io/camunda/zeebe/gateway/ResponseMapper.java#L81-L90) with metadata information and not with information from resources. This is why the test do not fail. What we want to do ?

[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:150]: Ah great. If it does not break user space there's no need to make further changes. Thanks for looking into it 👍

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentTest.java [hunk 3]
```diff
@@ -211,14 +210,9 @@ public void shouldCreateDeploymentResourceWithMultipleProcesses() {
...
-    assertThat(deployment.getValue().getResources())
-        .extracting(DeploymentResource::getResourceName)
+    assertThat(deployment.getValue().getProcessesMetadata())
+        .extracting(ProcessMetadataValue::getResourceName)
         .contains("process.bpmn", "process2.bpmn");
-
-    assertThat(deployment.getValue().getResources())
-        .extracting(DeploymentResource::getResource)
-        .contains(
-            Bpmn.convertToString(process).getBytes(), Bpmn.convertToString(process2).getBytes());
```
[the remark this answers is not in the document]
Comment (nicpuppa) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentTest.java]: Thank you for spotting it

**Review by korthout — APPROVED** [review]
Thanks for applying the requested changes @nicpuppa 💯 

👍 LGTM

💭 As this fixes a regression that is only present on `main` there's no need to backport it to other branches

bors merge

**Pull request merged**